// eslint-plugin-roblox-ts/out/rules/noPrivateIdentifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noPrivateIdentifier = exports.noPrivateIdentifierName = void 0;
const typescript_1 = require("typescript");
const rules_1 = require("../util/rules");
exports.noPrivateIdentifierName = "no-private-identifier";
exports.noPrivateIdentifier = rules_1.makeRule({
    name: exports.noPrivateIdentifierName,
    meta: {
        type: "problem",
        docs: {
            description: "Bans private identifiers from being used",
            category: "Possible Errors",
            recommended: "error",
            requiresTypeChecking: false,
        },
        messages: {
            privateIdentifierViolation: "Private identifiers are not supported!",
        },
        schema: [],
    },
    defaultOptions: [],
    create(context) {
        const service = rules_1.getParserServices(context);
        return {
            ClassProperty(node) {
                const tsNode = service.esTreeNodeToTSNodeMap.get(node.key);
                if (typescript_1.isPrivateIdentifier(tsNode)) {
                    context.report({
                        node: node,
                        messageId: "privateIdentifierViolation",
                    });
                }
            },
        };
    },
});
