// roblox-ts/out/TSTransformer/nodes/statements/transformVariableStatement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformVariableStatement = exports.transformVariableDeclarationList = exports.isVarDeclaration = exports.transformVariableDeclaration = exports.transformVariable = void 0;
const LuauAST_1 = __importDefault(require("../../../LuauAST"));
const diagnostics_1 = require("../../../Shared/diagnostics");
const assert_1 = require("../../../Shared/util/assert");
const getOrSetDefault_1 = require("../../../Shared/util/getOrSetDefault");
const DiagnosticService_1 = require("../../classes/DiagnosticService");
const transformArrayBindingPattern_1 = require("../binding/transformArrayBindingPattern");
const transformObjectBindingPattern_1 = require("../binding/transformObjectBindingPattern");
const transformExpression_1 = require("../expressions/transformExpression");
const transformIdentifier_1 = require("../expressions/transformIdentifier");
const transformInitializer_1 = require("../transformInitializer");
const isDefinedAsLet_1 = require("../../util/isDefinedAsLet");
const traversal_1 = require("../../util/traversal");
const types_1 = require("../../util/types");
const validateIdentifier_1 = require("../../util/validateIdentifier");
const typescript_1 = __importDefault(require("typescript"));
function checkVariableHoist(state, node, symbol) {
    if (state.isHoisted.get(symbol) !== undefined) {
        return;
    }
    const statement = (0, traversal_1.getAncestor)(node, typescript_1.default.isStatement);
    if (!statement) {
        return;
    }
    const caseClause = statement.parent;
    if (!typescript_1.default.isCaseClause(caseClause)) {
        return;
    }
    const caseBlock = caseClause.parent;
    const isUsedOutsideOfCaseClause = typescript_1.default.FindAllReferences.Core.eachSymbolReferenceInFile(node, state.typeChecker, node.getSourceFile(), token => {
        if (!(0, traversal_1.isAncestorOf)(caseClause, token)) {
            return true;
        }
    }, caseBlock) === true;
    if (isUsedOutsideOfCaseClause) {
        (0, getOrSetDefault_1.getOrSetDefault)(state.hoistsByStatement, statement.parent, () => new Array()).push(node);
        state.isHoisted.set(symbol, true);
    }
}
function transformVariable(state, identifier, right) {
    return state.capture(() => {
        (0, validateIdentifier_1.validateIdentifier)(state, identifier);
        const symbol = state.typeChecker.getSymbolAtLocation(identifier);
        (0, assert_1.assert)(symbol);
        if ((0, isDefinedAsLet_1.isDefinedAsLet)(state, symbol)) {
            const exportAccess = state.getModuleIdPropertyAccess(symbol);
            if (exportAccess) {
                if (right) {
                    state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
                        left: exportAccess,
                        operator: "=",
                        right,
                    }));
                }
                return exportAccess;
            }
        }
        const left = (0, transformIdentifier_1.transformIdentifierDefined)(state, identifier);
        checkVariableHoist(state, identifier, symbol);
        if (state.isHoisted.get(symbol) === true) {
            if (right) {
                state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, { left, operator: "=", right }));
            }
        }
        else {
            state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.VariableDeclaration, { left, right }));
        }
        return left;
    });
}
exports.transformVariable = transformVariable;
function transformLuaTupleDestructure(state, bindingPattern, value) {
    return state.capturePrereqs(() => {
        const ids = LuauAST_1.default.list.make();
        const statements = state.capturePrereqs(() => {
            for (const element of bindingPattern.elements) {
                if (typescript_1.default.isOmittedExpression(element)) {
                    LuauAST_1.default.list.push(ids, LuauAST_1.default.tempId());
                }
                else {
                    if (element.dotDotDotToken) {
                        DiagnosticService_1.DiagnosticService.addDiagnostic(diagnostics_1.errors.noSpreadDestructuring(element));
                        return;
                    }
                    if (typescript_1.default.isIdentifier(element.name)) {
                        (0, validateIdentifier_1.validateIdentifier)(state, element.name);
                        const id = (0, transformIdentifier_1.transformIdentifierDefined)(state, element.name);
                        LuauAST_1.default.list.push(ids, id);
                        if (element.initializer) {
                            state.prereq((0, transformInitializer_1.transformInitializer)(state, id, element.initializer));
                        }
                    }
                    else {
                        const id = LuauAST_1.default.tempId("binding");
                        LuauAST_1.default.list.push(ids, id);
                        if (element.initializer) {
                            state.prereq((0, transformInitializer_1.transformInitializer)(state, id, element.initializer));
                        }
                        if (typescript_1.default.isArrayBindingPattern(element.name)) {
                            (0, transformArrayBindingPattern_1.transformArrayBindingPattern)(state, element.name, id);
                        }
                        else {
                            (0, transformObjectBindingPattern_1.transformObjectBindingPattern)(state, element.name, id);
                        }
                    }
                }
            }
        });
        if (LuauAST_1.default.list.isEmpty(ids)) {
            LuauAST_1.default.list.push(ids, LuauAST_1.default.tempId());
        }
        state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.VariableDeclaration, { left: ids, right: value }));
        state.prereqList(statements);
    });
}
function transformVariableDeclaration(state, node) {
    const statements = LuauAST_1.default.list.make();
    let value;
    if (node.initializer) {
        LuauAST_1.default.list.pushList(statements, state.capturePrereqs(() => (value = (0, transformExpression_1.transformExpression)(state, node.initializer))));
    }
    if (typescript_1.default.isIdentifier(node.name)) {
        LuauAST_1.default.list.pushList(statements, transformVariable(state, node.name, value)[1]);
    }
    else {
        (0, assert_1.assert)(node.initializer && value);
        const name = node.name;
        if (typescript_1.default.isArrayBindingPattern(name)) {
            if (LuauAST_1.default.isCall(value) && (0, types_1.isLuaTupleType)(state)(state.getType(node.initializer))) {
                LuauAST_1.default.list.pushList(statements, transformLuaTupleDestructure(state, name, value));
            }
            else {
                LuauAST_1.default.list.pushList(statements, state.capturePrereqs(() => (0, transformArrayBindingPattern_1.transformArrayBindingPattern)(state, name, state.pushToVar(value, "binding"))));
            }
        }
        else {
            LuauAST_1.default.list.pushList(statements, state.capturePrereqs(() => (0, transformObjectBindingPattern_1.transformObjectBindingPattern)(state, name, state.pushToVar(value, "binding"))));
        }
    }
    return statements;
}
exports.transformVariableDeclaration = transformVariableDeclaration;
function isVarDeclaration(node) {
    return !(node.flags & typescript_1.default.NodeFlags.Const) && !(node.flags & typescript_1.default.NodeFlags.Let);
}
exports.isVarDeclaration = isVarDeclaration;
function transformVariableDeclarationList(state, node) {
    if (isVarDeclaration(node)) {
        DiagnosticService_1.DiagnosticService.addDiagnostic(diagnostics_1.errors.noVar(node));
    }
    const statements = LuauAST_1.default.list.make();
    for (const declaration of node.declarations) {
        const [variableStatements, prereqs] = state.capture(() => transformVariableDeclaration(state, declaration));
        LuauAST_1.default.list.pushList(statements, prereqs);
        LuauAST_1.default.list.pushList(statements, variableStatements);
    }
    return statements;
}
exports.transformVariableDeclarationList = transformVariableDeclarationList;
function transformVariableStatement(state, node) {
    return transformVariableDeclarationList(state, node.declarationList);
}
exports.transformVariableStatement = transformVariableStatement;
