// roblox-ts/out/TSTransformer/macros/constructorMacros.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONSTRUCTOR_MACROS = void 0;
const LuauAST_1 = __importDefault(require("../../LuauAST"));
const assert_1 = require("../../Shared/util/assert");
const transformExpression_1 = require("../nodes/expressions/transformExpression");
const ensureTransformOrder_1 = require("../util/ensureTransformOrder");
const typescript_1 = __importDefault(require("typescript"));
function wrapWeak(state, node, macro) {
    return LuauAST_1.default.call(LuauAST_1.default.globals.setmetatable, [
        macro(state, node),
        LuauAST_1.default.map([[LuauAST_1.default.strings.__mode, LuauAST_1.default.strings.k]]),
    ]);
}
const ArrayConstructor = (state, node) => {
    if (node.arguments && node.arguments.length > 0) {
        const args = (0, ensureTransformOrder_1.ensureTransformOrder)(state, node.arguments);
        return LuauAST_1.default.call(LuauAST_1.default.globals.table.create, args);
    }
    return LuauAST_1.default.array();
};
const SetConstructor = (state, node) => {
    if (!node.arguments || node.arguments.length === 0) {
        return LuauAST_1.default.set();
    }
    const arg = node.arguments[0];
    if (typescript_1.default.isArrayLiteralExpression(arg) && !arg.elements.some(typescript_1.default.isSpreadElement)) {
        return LuauAST_1.default.set((0, ensureTransformOrder_1.ensureTransformOrder)(state, arg.elements));
    }
    else {
        const id = state.pushToVar(LuauAST_1.default.set(), "set");
        const valueId = LuauAST_1.default.tempId("v");
        state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ForStatement, {
            ids: LuauAST_1.default.list.make(LuauAST_1.default.tempId(), valueId),
            expression: LuauAST_1.default.call(LuauAST_1.default.globals.ipairs, [(0, transformExpression_1.transformExpression)(state, arg)]),
            statements: LuauAST_1.default.list.make(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
                left: LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ComputedIndexExpression, {
                    expression: id,
                    index: valueId,
                }),
                operator: "=",
                right: LuauAST_1.default.bool(true),
            })),
        }));
        return id;
    }
};
const MapConstructor = (state, node) => {
    if (!node.arguments || node.arguments.length === 0) {
        return LuauAST_1.default.map();
    }
    const arg = node.arguments[0];
    const transformed = (0, transformExpression_1.transformExpression)(state, arg);
    if (LuauAST_1.default.isArray(transformed) && LuauAST_1.default.list.every(transformed.members, member => LuauAST_1.default.isArray(member))) {
        const elements = LuauAST_1.default.list.toArray(transformed.members).map(e => {
            (0, assert_1.assert)(LuauAST_1.default.isArray(e) && LuauAST_1.default.list.isNonEmpty(e.members));
            return [e.members.head.value, e.members.head.next.value];
        });
        return LuauAST_1.default.map(elements);
    }
    else {
        const id = state.pushToVar(LuauAST_1.default.map(), "map");
        const valueId = LuauAST_1.default.tempId("v");
        state.prereq(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ForStatement, {
            ids: LuauAST_1.default.list.make(LuauAST_1.default.tempId(), valueId),
            expression: LuauAST_1.default.call(LuauAST_1.default.globals.ipairs, [transformed]),
            statements: LuauAST_1.default.list.make(LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.Assignment, {
                left: LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ComputedIndexExpression, {
                    expression: id,
                    index: LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ComputedIndexExpression, {
                        expression: valueId,
                        index: LuauAST_1.default.number(1),
                    }),
                }),
                operator: "=",
                right: LuauAST_1.default.create(LuauAST_1.default.SyntaxKind.ComputedIndexExpression, {
                    expression: valueId,
                    index: LuauAST_1.default.number(2),
                }),
            })),
        }));
        return id;
    }
};
exports.CONSTRUCTOR_MACROS = {
    ArrayConstructor,
    SetConstructor,
    MapConstructor,
    WeakSetConstructor: (state, node) => wrapWeak(state, node, SetConstructor),
    WeakMapConstructor: (state, node) => wrapWeak(state, node, MapConstructor),
    ReadonlyMapConstructor: MapConstructor,
    ReadonlySetConstructor: SetConstructor,
};
